Refactor 2: Refactor Harder

jtobin opened the pull request:
Per @Fang-'s comment on #1, this is (also) a big one.

(Context: I approached this library as something I inherited, in order to jack it into Bridge.  I obviously haven't tried to be incremental at all here, instead opting to replace one reconstitution with another.)

This PR restructures the library again, mostly keeping @Fang-'s broad changes but exposing them via a purely-functional API where state passing is explicit.  Additionally I've changed the internal structure of the library somewhat -- functions in `/internal` modules are intended to maintain a one-to-one mapping with the actual contract APIs (to stay utterly "close to the contracts"), while other modules are free to expose their own implementations of these (only notable at present via the type-generic `ships` API -- functions in which can often take ship objects, and avoid hitting the network).

For the user, the manual state-passing (typically of a `contracts` object, or occasionally `web3` in the case of transactions) is the biggest change.  Previously we would have:

```javascript
constitution.createGalaxy(galaxy, account)
```

whereas now we need to pass the contract (actually, all relevant contracts) explicitly as the first argument:

```javascript
createGalaxy(contracts, galaxy, account)
```

Similarly, at no point is there the concept of the "current account" -- accounts must be passed explicitly, where required.  The `from` field of transactions is derived from the private key used to sign them -- not from the (implicit) current state of web3 objects.

This all makes the library easier to modify (there is never the need to assume the existence of a web3 or contract instance in library code, for example), but to the user, it otherwise feels much the same.  Pass in state, get result.  There are also other changes: additional supported contract functions, bug fixes, deriving addresses from HD nodes directly, making more use of `ethereumjs-util` functions, handling private keys as buffers, etc.

I've begun to use this code in my work on Bridge and find it plays very well with things in that setting.  Someone should let me know if there are any obvious problems with it as they see it being used elsewhere! 😁

Commit:
Begin rewrite of constitution-js

Currently includes all Constitution contract functionality, but is
untested save for the generic transaction generation and signing logic.
Pool contract logic is entirely missing, as are data validation
functions.

General philosophy here is to wrap around all of the contracts'
functions in a consistent way.
Generating a transaction (offline, synchronous) is decoupled from
checking if a transaction will succeed (online, asynchronous).
While all the relevant logic is provided, it is up to the UI to call
upon this logic when doing input and transaction validation, prior to
transaction generation.

The transaction check functions, in cases where a transaction is
concluded to not be successful, return a reason alongside the false
result. The UI can use this to inform the user about why their desired
transaction is rejected.

Some data retrieval functions are provided, but they do not wrap around
all of the getter provided in the Ships contract. Instead, when using
this library it is encouraged to use the broader getShip() (or
getPoll()) function, rather than more granular reads, to limit the
amount of network traffic (and thus UX lag) that has to be done.

Commit:
Fix minor logic errors

Commit:
Include contract ABIs

Commit:
Split contract functions out into separate files

Also includes minor changes to logic here and there.

Commit:
Add functions for interfacing with Pool contracts

Commit:
Add input validation utilities

Commit:
Add tests, fix all the issues they found

Commit:
Do cleanup, apply small fixes and additions

Commit:
Remove unused file

Commit:
Document different modules, update readme to include overview

Commit:
Update package.json to conform to internal standard

Commit:
Remove outdated examples from readme

vvisigoth committed:
added websockets to package.json

jtobin committed:
Ignore vim swap files, reference stuff, tags file.

jtobin committed:
Remove old test code.

jtobin committed:
General restructuring.

Per Mark's commentary re: the audit of the old constitution-js code, the
goals of his rewrite were:

* Stay close to the contracts
* Minimise node calls
* Don't take work out of the UI/client's hands
* But send useful error messages

I've restructured things such that, primarily:

* The library is purely-functional (to the extent that is possible in
  JS) and exposes a purely-functional API.  In particular: no unmanaged
  state.

* Code in 'internal' modules consists of simple, one-to-one mappings with
  Solidity contract code -- nothing fancy goes on here.  These are
  designated 'internal' as users need not deal with them directly.

* Code in 'ships', 'constitution', etc. consists of the higher-level,
  user-facing API.  These functions are generic, and can make use of local
  data to save on network calls where applicable.  At present this is only
  applicable to the ships API, but we may make other use of this
  structure in the future.

  Additionally, these modules contain functions that do not necessarily
  map directly to contract code -- verifications and assertions, for
  example.

* Private keys are never handled as raw strings, old dependencies are
  gutted, plus other misc. improvements.

jtobin committed:
Add docs (via JSDoc).

jtobin committed:
Misc bug fixes.

* Get rid of cached node_modules (surely I didn't commit those..).
* Fix botched find-and-replace on 'protoTx'.
* Fix naming mismatches in ships API.
* Do not pass last parameter of safeTransferFrom (let Solidity handle).
* Fix pool.getBalance and pool.deposit impl. bugs.

jtobin committed:
Implement pool.withdraw, withdrawAny.

jtobin committed:
Docs fix-ups.

Adds JSDocs for the ships and check APIs, and tidies up the JSDocs for
the constitution API.

I also encounterd a bizarre and very Javascript-y error when assigning
functions to module.exports directly (i.e. at the binding site) in
constitution.js.  Some functions failed to be exported.  I gave all the
re-exported functions const declarations instead, and this fixed it up.

Eventually I'll probably ditch the module.exports.foo syntax entirely.

jtobin committed:
Use ethereumjs-util in tests.

(We don't need our own utils module here.)

jtobin committed:
Remove old docs.

jtobin committed:
Fix up JSDoc syntax.

jtobin committed:
Refactor modules.

Split the account and contract initialisation functionality into their
own modules.

Also, do not export the non-confirming 'sendTransaction' function for
unsigned transactions.

jtobin committed:
More docs updates.

jtobin committed:
Merge branch 'jt-reconstitute'

Fang- commented on contracts.js:
```diff
@@ -0,0 +1,44 @@
+ *   names.
+ * @return {Object} The initialised contracts.
+ */
+function initContracts(web3, addresses) {
```
What happened to the version of this function that took just the Ships address (and the one that provided the (known and static after launch) address for you) and read out all the other addresses from there? I think that's a pretty nice affordance, makes sure dependents don't have to keep addresses up-to-date themselves.

jtobin replied:
It might have just fallen out in the refactor.  But in general I've probably been thinking of this library in network-generic fashion, so that clients must always supply their own addresses.  Just thinking in terms of (abstract) dependencies: I'd like to pin a version of the *contracts* as a dependency, rather than a particular network.

(This may be overly general on my part -- there is only one Ethereum mainnet, after all. 😁)

Fang- replied:
Not only is there just one Ethereum mainnet, there is (or, will be) exactly one collection of contracts we call the Constitution. Of course for testnet cases you still want to be able to specify manually. I'm fine with pinning for the real-world case though.

Fang- reviewed (commented):
This is a big one! Majority of it is cleaning up `node_modules`, which for some reason was still in this repo...

On the one hand, being able to pass in your own state (web3, contracts) makes a lot of sense, and seems really useful if you actually know what you're dealing with.  
On the other hand, my idea behind containing all that within the library was that clients didn't *need* to know anything about web3 to be able to interface with the contracts.  
Of course, currently we're the only ones who are using the library, and we do know what it's all about, but I think it's nice to keep this accessible to eager community developers.

That said, I don't know how well that would fit in with this new version of the library. We might want to do the same for web3 as we're currently doing for the contracts: providing a function to retrieve a usable "state" object.  
It feels a bit weird still to ask the user to pass around an object they're not required to know anything about, but I'm also wildly unfamiliar with functional Javascript, so maybe it's not all that weird...

In that vein, I don't feel like I can make an informed call on whether or not this is good JS style or not. Overall I would say this looks pretty good, if a bit redundant in places. A quick review by any of @gavinpatkinson @c-johnson @vvisigoth would be great to have.

Also, if we're doing auto-generated docs, we might want to include instructions on how to regenerate them in the readme.

`check.js` has some weird syntax highlighting going on, I wonder what's up with that.

Fang- commented:
Actually, re: regenerating docs, might be nice to have that done by CI if that's not too difficult.

g-a-v-i-n commented on contracts.js:
```diff
@@ -0,0 +1,44 @@
+  return new web3.eth.Contract(abis.polls, address);
+}
+
+function newPool(web3, address) {
```
Nitpick alert, but one other way to write these is:
```js
const newPool = (web3, address) => new web3.eth.Contract(abis.pool, address);
```

Arrow function without {} imply the return keyword.

jtobin replied:
Right.  I prefer the latter, as well.  I will probably take a pass over the library to standardise the style (though I would be inclined to make another PR for that).

g-a-v-i-n commented on internal/polls.js:
```diff
@@ -0,0 +1,29 @@
+}
+
+module.exports.constitutionHasAchievedMajority = (contracts, proposal) => {
+  return contracts.polls.methods.constitutionHasAchievedMajority(proposal).call();
```
Why do we need .call() here?

Fang- replied:
That's a web3 thing, it makes the read/call you defined actually happen.

g-a-v-i-n replied:
I see, thanks

g-a-v-i-n commented on check.js:
```diff
@@ -0,0 +1,459 @@
+ * @return {Promise<Bool>} True if so, false otherwise.
+ */
+async function canConfigureKeys(contracts, ship, address) {
+  res = { result: false };
```
Does this variable definition work without `let`?

jtobin replied:
It.. apparently does? 😕  That function gets called a couple of times in the test suite, and the tests are passing.  I'll fix it, though.

g-a-v-i-n commented on check.js:
```diff
@@ -0,0 +1,459 @@
+async function canConfigureKeys(contracts, ship, address) {
+  res = { result: false };
+  // must be able to manage ship
+  if (!await ships.canManage(contracts, ship, address))
```
Generally I prefer explicit typechecking on this sort of thing, like:

```js
if (isFalse(ships.canManage(contracts, ship, address))) {}
```

g-a-v-i-n commented on check.js:
```diff
@@ -0,0 +1,459 @@
+ * @return {Promise<Bool>} True if so, false otherwise.
+ */
+async function hasOwner(contracts, ship) {
+  if (typeof ship === 'object') {
```
I would consider using something like below. Typechecking objects in JS is weird. If you pass an array as the `ship` param it will pass through this test.

```js
const isObject = any => any && typeof any === 'object' && a.constructor === Object
```

jtobin replied:
This (with typo fixed) actually breaks all the assertion functions, since they're also expected to work on promises (so, the last check doesn't necessarily hold).  It took me a little bit of time to figure out WTF was going on.

I think the existing check is ok for now, since the aim of this code is to expose a generic API, rather than prevent bogus inputs *per se*.

That said, branching on types generally gives me the willies, no matter what language I'm in, so I'm inclined to export an alternative pure ships API where passing an ship object (rather than integer) is done explicitly.  One could require `cjs.ships` or `cjs.ships.pure` separately, for example, depending on what one is interested in doing.  But I think figuring out the details of that is something for down the road.

Fang- commented on check.js:
```diff
@@ -0,0 +1,459 @@
+ * @param {String} address - Target address.
+ * @return {Promise<Bool>} True if so, false otherwise.
+ */
+async function canConfigureKeys(contracts, ship, address) {
```
Any reason you're not just using `checkActiveShipManager` here, as you do in other places, and as #1 does?

jtobin replied:
Huh, nope.  I think I was on the track of thinking that `checkActiveShipManager` should be replaced by `canConfigureKeys`, or vice versa, or merged together as something else (since they check for the same thing), but forgot to follow up on it.  Will fix.

Both boil down to `canManageAndShipIsActive`, and so in theory could be replaced by that.  But that's a mouthful. 😄

jtobin replied:
Hmm.  I fixed these, though now that I think on it, it sort of goes against the style of the API.

I'll rename to `canConfigureKeys`, simply because there exists a `configureKeys` function in the API.

Fang- replied:
Woah, no, that's the opposite of what I wanted here lol.

`canManageAndShipIsActive` is a specific check about the state it describes. Other checks should use this as a building block for their often more complex checks.  
`canConfigureKeys` is a function check about the ability to call that specific function. It just so happens that the only check that does is `canManageAndShipIsActive`, and that's totally fine.

Surely that makes more sense than having function checks depend on function checks? The above also more closely matches to the contracts themselves, something I should double-check wasn't lost-in-translation for other checks either.

Fang- replied:
So, to be clear on my vision here:

- There should be a `canX` check for every function the Constitution contract exposes,
- There should be helper functions for parts of those checks in the same way the Constitution contract has them.

That way we stay nice and close to the contracts, and don't force the user to reason about what it takes to be able to successfully call a specific function. Sensical y/n?

jtobin replied:
Ahhh right, gotcha.  I missed that you were matching those helper functions with the actual assertions in the contracts -- I thought it was a redundant internal thing that I could safely eliminate.  Ok, I'll put `activeShipManager` back in.

I think this is the only one that actually got *changed* -- there are another few assertions around polls/voting that I removed for the time being, just because I wasn't using them in bridge, but I can put them back in order to be consistent.

Fang- replied:
Yes, please keep feature/logic parity with `reconstitute`. (:

jtobin committed:
Fix lack-of-let.

jtobin committed:
Replace active ship manager checks.

Replaces canConfigureKeys and checkActiveShipManager (which do the same
thing) with canManageAndShipIsActive.

jtobin committed:
s/canManageAndShipIsActive/canConfigureKeys

There actually exists a configureKeys function, so this name fits the
API a little better.

jtobin commented:
@Fang- I understand your point re: hiding the web3 details from the user.  But I think anyone pulling in `constitution-js` will probably already be pulling in `web3` anyway, since they will probably want to do something or other that we haven't provided for them in the library.  I think the present API works well *with* web3, and doesn't try to do what web3 can do just as well or better.

To your point on CI: definitely.  I'm planning to get a basic-but-useful Travis config set up when I get a chance.

jtobin committed:
Replace the checkActiveShipManager check.

Per @Fang-, the idea behind this function is that it duplicates an
assertion made in the constitution contract itself.

jtobin committed:
Add poll functions, poll tests.

jtobin committed:
Add poll tests.

jtobin committed:
Misc uncommitted stuff.

jtobin committed:
Add missing pool stuff.

jtobin committed:
Add partial contract initialisation.

jtobin committed:
Add polls documentation.

jtobin committed:
Pool documentation patches.

jtobin committed:
Add nyc to dev dependencies.

jtobin commented:
Ok -- going to merge this one now so I can make future commits a little more sanely-sized.
